```bash
#!/bin/bash
#SBATCH --job-name=TASK
#SBATCH --output=TASK.out
#SBATCH --error=TASK.err
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=1
#SBATCH --cpus-per-task=12
#SBATCH --mem=25G
#SBATCH --partition=base
#SBATCH --time=5:00:00
#SBATCH --reservation=biol217

#load necessary modules
module load gcc12-env/12.1.0
module load micromamba
eval "$(micromamba shell hook --shell=bash)"
export MAMBA_ROOT_PREFIX=$WORK/.micromamba

cd $WORK

micromamba activate .micromamba/envs/00_anvio/
```

We performed binning and quality assesment for our assembly previous day. Today we will work on bin refinement and chimera detection.

## Bin Refinement

Bin refinement is essential to improve the quality of metagenome-assembled genomes (MAGs). Even with automated binning tools like MetaBAT2, bins may contain contaminations, chimeric sequences, or low-quality bins. Refining bins ensures higher-quality MAGs that are more representative of individual genomes.

For this step, we focused specifically on archaeal bins. So we Firstly need to refine our Archea bins from others. I used `anvi-summarize` to generate a summary of all bins. This summary provided critical information, including:
- The taxonomy of each bin
- Completeness and contamination levels

```bash
# Summary Generation
module load gcc12-env/12.1.0
module load micromamba
cd $WORK
micromamba activate .micromamba/envs/00_anvio/

cd /work_beegfs/sunam227/metagenomics/0_raw_reads/clean
anvi-summarize -p ./mergedprof/PROFILE.db -c contigs.db --list-collections

cd /work_beegfs/sunam227/metagenomics/0_raw_reads/clean
anvi-summarize -c contigs.db -p ./mergedprof/PROFILE.db -C METABAT2 -o ./SUMMARY_METABAT2 --just-do-it
```

From the summary, I identified which bins were archaeal. Since MetaBAT2 produced the better results in the initial binning step, I continued using the bins generated by MetaBAT2 for further refinement.

### !!!SUMMARY ISTATISTIK KOY!!!

Using these Archeal bins we will perform a Chimera detection in MAGs. But first we need to refine these bins from non-archeal ones. Fallowing Archaeal bins were copied into a separate directory for focused analysis.
- METABAT__8
- METABAT__10
- METABAT__36

```bash
# Archeal Bin Refinment
cd /work_beegfs/sunam227/metagenomics/0_raw_reads/clean/SUMMARY_METABAT2/bin_by_bin

mkdir ../../ARCHAEA_BIN_REFINEMENT

cp ./METABAT__10/*.fa ../../ARCHAEA_BIN_REFINEMENT
cp ./METABAT__36/*.fa ../../ARCHAEA_BIN_REFINEMENT
cp ./METABAT__8/*.fa ../../ARCHAEA_BIN_REFINEMENT
```

## Chimera Detection in MAGs

A chimeric genome is an assembly that combines sequences from multiple, unrelated organisms into a single contig or bin. This can occur during assembly due to:
- Errors in resolving repetitive regions
- Incorrect scaffolding of reads from closely related species
- Misbinning during the binning process

Chimeric genomes can distort downstream analyses, such as phylogenetic classification, functional annotation, or metabolic modeling, making their detection and resolution crucial.

I used `GUNC` (Genome UNClutterer) to evaluate each archaeal bin for chimerism and contamination.

```bash
module load gcc12-env/12.1.0
module load micromamba/1.3.1
micromamba activate 00_gunc

cd /work_beegfs/sunam227/metagenomics/0_raw_reads/clean/SUMMARY_METABAT2/bin_by_bin/ARCHAEA_BIN_REFINEMENT

for i in *.fa; do mkdir /work_beegfs/sunam227/metagenomics/0_raw_reads/06_gunc/"$i"_out; done

for i in *.fa; do
  gunc run -i "$i" -r $WORK/databases/gunc/gunc_db_progenomes2.1.dmnd --out_dir /work_beegfs/sunam227/metagenomics/0_raw_reads/06_gunc/"$i"_out --threads 12 --detailed_output
done


cd /work_beegfs/sunam227/metagenomics/0_raw_reads/06_gunc/METABAT__8-contigs.fa_out
gunc plot -d ./diamond_output/METABAT__8-contigs.diamond.progenomes_2.1.out -g ./gene_calls/gene_counts.json

cd /work_beegfs/sunam227/metagenomics/0_raw_reads/06_gunc/METABAT__10-contigs.fa_out
gunc plot -d ./diamond_output/METABAT__10-contigs.diamond.progenomes_2.1.out -g ./gene_calls/gene_counts.json

cd /work_beegfs/sunam227/metagenomics/0_raw_reads/06_gunc/METABAT__36-contigs.fa_out
gunc plot -d ./diamond_output/METABAT__36-contigs.diamond.progenomes_2.1.out -g ./gene_calls/gene_counts.json
```

`GUNC` provides a `CSS score` (Chimeric Sequence Score) for each bin.
- Low CSS scores indicate low levels of chimerism and contamination.
- High CSS scores suggest a bin is likely chimeric or highly contaminated.

`GUNC` also detects redundancy, which refers to the presence of duplicate marker genes or repetitive sequences within a bin.

A bin with a high CSS score and significant redundancy likely requires manual refinement or exclusion from further analyses.



Here, we can see some result of our GUNC analysis

### !!!DAHA Ã‡OK EKLE!!!

### METABAT__8
![image](./resources/Contigs_8.png)
### METABAT__10
![image](./resources/Contigs_10.png)
### METABAT_36
![image](./resources/Contigs_36.png)

## Manual Refinment of Bins

In case needed, a manual refinment can be used if there are not too much of bins. For manual refinement, bins should be selected based on completeness (>70%) and low contamination. High-quality bins were prioritized for manual inspection to save time and effort.

Bins can manually refined by:
1. Sorting bins by GC content and coverage: Sorting helps identify outliers, as bins from the same organism typically have consistent GC content and coverage. Outliers may indicate contamination or chimerism.
2. Removing problematic contigs: Contigs that showed abnormal GC content, low coverage, or conflicting taxonomy can manually removed from the bins.
3. Evaluating taxonomy: Taxonomic assignments were cross-checked to ensure that all contigs in a bin were taxonomically consistent. Contigs with mismatched taxonomy can flagged for removal.

## Questions
1- Does the quality of your Archea improved?
- No, they remain the same. There might be a problem occured during bin refinement.